samtranslator/model/sam_resources.py

```diff
@@ -2194,3 +2201,106 @@ def _construct_appsync_schema(self, api_id: Intrinsicable[str]) -> GraphQLSchema
         schema.DefinitionS3Location = self.SchemaUri
 
         return schema
+
+
+class SamGraphQLDataSource(SamResourceMacro):
+    """SAM GraphQL Data Source Macro (WIP)."""
+
+    # TODO: This currently only supports DataSource with a defined ServiceRoleArn.
+    #       The proposed connector change will come in a follow up PR.
+    resource_type = "AWS::Serverless::GraphQLDataSource"
+    property_types = {
+        "ApiId": PassThroughProperty(True),
+        "Name": Property(False, IS_STR),
+        "Type": Property(True, IS_STR),
+        "Description": PassThroughProperty(False),
+        "ServiceRoleArn": Property(False, IS_STR),
+        "DynamoDBConfig": Property(False, IS_DICT),
+    }
+
+    ApiId: PassThrough
+    Type: str
+    DynamoDBConfig: Optional[Dict[str, Any]]
+    Name: Optional[str]
+    Description: Optional[PassThrough]
+    ServiceRoleArn: Optional[str]
+
+    @cw_timer
+    def to_cloudformation(self, **kwargs: Any) -> List[Resource]:
+        appsync_datasource = self._construct_appsync_datasource()
+        resources: List[Resource] = [appsync_datasource]
+
+        return resources
+
+    def _parse_deltasync_properties(self, deltasync_properties: Dict[str, Any]) -> DeltaSyncConfigType:
+        sam_expect(
+            deltasync_properties.get("BaseTableTTL"), self.logical_id, "DynamoDBConfig.DeltaSyncConfig.BaseTableTTL"
+        ).to_be_a_string()
+
+        sam_expect(
+            deltasync_properties.get("DeltaSyncTableName"),
+            self.logical_id,
+            "DynamoDBConfig.DeltaSyncConfig.DeltaSyncTableName",
+        ).to_be_a_string()
+
+        sam_expect(
+            deltasync_properties.get("DeltaSyncTableTTL"),
+            self.logical_id,
+            "DynamoDBConfig.DeltaSyncConfig.DeltaSyncTableTTL",
+        ).to_be_a_string()
+
+        return cast(DeltaSyncConfigType, deltasync_properties)
+
+    def _parse_dynamodb_datasource(self, ddb_properties: Dict[str, Any]) -> DynamoDBConfigType:
+        ddb_config: Dict[str, Any] = {}
```

Comment:
IMO rather than doing this and `cast`ing, we should build the `TypedDict` directly.

Reply:
Like just change the typing on `ddb_config` to `DynamoDBConfigType` and return without `cast`?

Reply:
Maybe, but not sure if `mypy` is smart enough since it's also kind of an assertion. I was thinking something like:

```
return DynamoDBConfig(
    Versioned=...,
    ...
)
```

Reply:
I thought of doing something like that, but optional properties like `Versioned` aren't always defined, so creating the dict in just 1 call isn't possible I think.

If I was to set the value to `None` as a default, it shows up as `null` in the transformed template as well which is obviously not appropriate.

Reply:
You can still assign stuff to typed variables before creating the final `TypedDict`.

> If I was to set the value to None as a default, it shows up as null in the transformed template as well which is obviously not appropriate.

How is that related to `TypedDict`?

Reply:
Offline discussion: The `TypedDict` is partial, so we cant make one call that suits all combinations. I will change type of `ddb_config` to `DynamoDBConfigType` and remove the cast.